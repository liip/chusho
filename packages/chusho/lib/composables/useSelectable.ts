import {
  ComputedRef,
  Ref,
  computed,
  getCurrentInstance,
  ref,
  watch,
} from 'vue';

export type SelectedItemId = string | number;

export interface SelectedItem<DataT = null> {
  id: SelectedItemId;
  data?: DataT;
}

export interface UseSelectable<ItemDataT = null> {
  selectedItemId: ComputedRef<SelectedItemId | null>;
  selectedItemIndex: ComputedRef<number | null>;
  selectedItem: ComputedRef<SelectedItem<ItemDataT> | null>;
  items: ComputedRef<SelectedItem<ItemDataT>[]>;
  setSelectedItem: (id: SelectedItemId) => void;
  addItem: (id: SelectedItemId, data?: Ref<ItemDataT>) => void;
  removeItem: (id: SelectedItemId) => void;
}

/**
 * Generic logic for a single selected item in a list of items, optionally binded to a prop on the parent component.
 */
export default function useSelectable<ItemDataT = null>(
  initialValue: SelectedItemId | null = null,
  propName: string | null = null
): UseSelectable<ItemDataT> {
  const vm = getCurrentInstance();
  const selectedItemId = ref<SelectedItemId | null>(initialValue);
  const items = ref([]) as Ref<SelectedItem<ItemDataT>[]>;

  function setSelectedItem(id: SelectedItemId) {
    selectedItemId.value = id;
    if (propName && vm) {
      vm.emit(`update:${propName}`, selectedItemId.value);
    }
  }

  function addItem(id: SelectedItemId, data?: Ref<ItemDataT>) {
    const item: SelectedItem<ItemDataT> = {
      id,
    };
    if (data) {
      item.data = data.value;
    }
    items.value.push(item);
  }

  function removeItem(id: SelectedItemId) {
    if (selectedItemId.value === id) {
      selectedItemId.value = items.value[0].id;
    }

    items.value.splice(
      items.value.findIndex((item) => item.id === id),
      1
    );
  }

  if (propName) {
    watch(
      () => vm?.props?.[propName],
      (val, oldVal) => {
        if (
          val !== oldVal &&
          (typeof val === 'string' || typeof val === 'number')
        ) {
          selectedItemId.value = val;
        }
      }
    );
  }

  return {
    selectedItemId: computed(() => selectedItemId.value),
    selectedItemIndex: computed(() =>
      items.value.findIndex((item) => item.id === selectedItemId.value)
    ),
    selectedItem: computed(
      () => items.value.find((item) => item.id === selectedItemId.value) ?? null
    ),
    items: computed(() => items.value),
    setSelectedItem,
    addItem,
    removeItem,
  };
}
