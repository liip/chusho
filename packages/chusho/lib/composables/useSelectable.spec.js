import { mount } from '@vue/test-utils';

import useSelectable from './useSelectable';

describe('useSelectable', () => {
  it('initialize with default value', () => {
    const selectable = useSelectable('1');

    expect(selectable.selectedItemId.value).toBe('1');
  });

  it('emits when there’s a component instance, a prop name and the value changes', () => {
    const wrapper = mount({
      emits: ['update:selectable'],
      setup() {
        return {
          selectable: useSelectable('1', 'selectable'),
        };
      },
      mounted() {
        this.selectable.setSelectedItem('2');
      },
      template: '<div></div>',
    });

    expect(wrapper.emitted()).toEqual({ 'update:selectable': [['2']] });
  });

  it('updates the selectedItem when there’s a component instance and the prop value changes', async () => {
    const wrapper = mount(
      {
        props: ['modelValue'],
        setup(props) {
          return {
            selectable: useSelectable(props.modelValue, 'modelValue'),
          };
        },
        template: '<div></div>',
      },
      {
        props: {
          modelValue: '1',
        },
      }
    );

    expect(wrapper.vm.selectable.selectedItemId.value).toEqual('1');
    await wrapper.setProps({ modelValue: '2' });
    expect(wrapper.vm.selectable.selectedItemId.value).toEqual('2');
    // It should do nothing if the value didn’t actually change
    await wrapper.setProps({ modelValue: '2' });
    expect(wrapper.vm.selectable.selectedItemId.value).toEqual('2');
  });

  it('setSelectedItem changes selectedItem', () => {
    const selectable = useSelectable('1');

    selectable.setSelectedItem('2');

    expect(selectable.selectedItemId.value).toBe('2');
  });

  it('addItem adds item to the list of items', () => {
    const selectable = useSelectable();

    selectable.addItem('1');
    selectable.addItem('2');
    selectable.addItem('3');

    expect(selectable.items.value).toMatchSnapshot();
  });

  it('removeItem removes item from the list of items', () => {
    const selectable = useSelectable();

    selectable.addItem('1');
    selectable.addItem('2');
    selectable.addItem('3');

    expect(selectable.items.value).toMatchSnapshot();

    selectable.removeItem('2');
    expect(selectable.items.value).toMatchSnapshot();

    selectable.removeItem('3');
    expect(selectable.items.value).toMatchSnapshot();

    selectable.removeItem('1');
    expect(selectable.items.value).toMatchSnapshot();
  });

  it('provides the selected item, its ID and index', () => {
    const selectable = useSelectable('1');

    selectable.addItem('1');

    expect(selectable.selectedItem.value).toEqual({ id: '1' });
    expect(selectable.selectedItemId.value).toBe('1');
    expect(selectable.selectedItemIndex.value).toBe(0);
  });
});
