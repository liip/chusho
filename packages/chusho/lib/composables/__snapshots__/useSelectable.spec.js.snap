// Vitest Snapshot v1

exports[`useSelectable > addItem adds item to the list of items 1`] = `
[
  {
    "id": "1",
  },
  {
    "id": "2",
  },
  {
    "id": "3",
  },
]
`;

exports[`useSelectable > removeItem removes item from the list of items 1`] = `
[
  {
    "id": "1",
  },
  {
    "id": "2",
  },
  {
    "id": "3",
  },
]
`;

exports[`useSelectable > removeItem removes item from the list of items 2`] = `
[
  {
    "id": "1",
  },
  {
    "id": "3",
  },
]
`;

exports[`useSelectable > removeItem removes item from the list of items 3`] = `
[
  {
    "id": "1",
  },
]
`;

exports[`useSelectable > removeItem removes item from the list of items 4`] = `[]`;
